scalyr_agent/builtin_monitors/linux_process_metrics.py

```diff
@@ -619,163 +741,336 @@ class ProcessMonitor(ScalyrMonitor):
+        return matching_pids
 
-        @return: The process id of the matching process, or None
-        @rtype: int or None
+    def should_poll_for_processes(self):
         """
-        sub_proc = None
-        if self.__commandline_matcher is not None:
-            try:
-                # Spawn a process to run ps and match on the command line.  We only output two
-                # fields from ps.. the pid and command.
-                sub_proc = Popen(['ps', 'ax', '-o', 'pid,command'],
-                                 shell=False, stdout=PIPE)
-
-                sub_proc.stdout.readline()
-                matching_pids = []
-                for line in sub_proc.stdout:
-                    line = line.strip()
-                    if line.find(' ') > 0:
-                        pid = int(line[:line.find(' ')])
-                        line = line[(line.find(' ') + 1):]
-                        if re.search(self.__commandline_matcher, line) is not None:
-                            matching_pids.append( pid )
+        We should poll for new processes in the following case:
 
-                total_pids = len( matching_pids )
-                if total_pids == 0:
-                    return None
-                elif total_pids == 1:
-                    return matching_pids[0]
+        * We do not currently have a process that we matched.
+        * This is a single process case, and the old process died.
+        * This is a multi-process case, and at least one process is alive and a certain time has expired.
+        :return: True if there is a need to poll, False otherwise
+        """
 
-                # if we are here then multiple command lines matched.  First check to
-                # see if self.__pid is in the list and if so return that.  Otherwise
-                # return the first element in the list
-                if self.__pid in matching_pids:
-                    self._logger.warning( "Multiple processes match the command '%s'.  Returning existing pid." % self.__commandline_matcher, limit_once_per_x_secs=300, limit_key='linux-process-monitor-existing-pid' )
-                    return self.__pid
+        if not self.__pids:
+            # no processes are currently being monitored, poll again
+            return True
+
+        if not self.__aggregate_multiple_processes:
+            # single process case
+            active_matched_pids = self.get_active_matched_pids(check_poll=False)
+            if not active_matched_pids:
+                # no matched processes, try again after sometime
+                return False
+            if active_matched_pids[0] not in self.__pids:
+                # matched process, but not running, old process maybe died
+                return True
+        else:
+            # multi-process case, poll every `poll_interval` seconds to
+            # see if there are more processes to match
+            if (
+                    not self.LAST_POLLED.get(self.__last_polled_env_key) or
```

Comment:
This is a bit complex.  You should simplify this by just checking if `poll_interval` plus  some `__last_time_check` defined for this instance is less than the current time.  If it is, then it is time to poll.